x-pack/plugins/maps/public/layers/sources/source.ts

```diff
@@ -0,0 +1,195 @@
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+/* eslint-disable @typescript-eslint/consistent-type-definitions */
+
+import { ReactElement } from 'react';
+
+// eslint-disable-next-line @kbn/eslint/no-restricted-paths
+// @ts-ignore
+import { copyPersistentState } from '../../reducers/util';
+
+import { LayerDescriptor, SourceDescriptor } from '../../../common/descriptor_types';
+import { ILayer } from '../layer';
+import { IField } from '../fields/field';
+
+export type ImmutableSourceProperty = {
+  label: string;
+  value: string;
+};
+
+export type Attribution = {
+  url: string;
+  label: string;
+};
+
+export type PreIndexedShape = {
+  index: string;
+  id: string | number;
+  path: string;
+};
+
+export type FieldFormatter = (value: string | number | null | undefined | boolean) => string;
+
+export interface ISource {
+  createDefaultLayer(options?: LayerDescriptor): ILayer;
+  destroy(): void;
+  getDisplayName(): Promise<string>;
+  getInspectorAdapters(): object | undefined;
+  isFieldAware(): boolean;
+  isFilterByMapBounds(): boolean;
+  isGeoGridPrecisionAware(): boolean;
+  isQueryAware(): boolean;
+  isRefreshTimerAware(): boolean;
+  isTimeAware(): Promise<boolean>;
+  getImmutableProperties(): Promise<ImmutableSourceProperty[]>;
+  getAttributions(): Promise<Attribution[]>;
+  isESSource(): boolean;
+  renderSourceSettingsEditor({ onChange }: { onChange: () => void }): ReactElement<any> | null;
+  supportsFitToBounds(): Promise<boolean>;
+  isJoinable(): boolean;
+  cloneDescriptor(): SourceDescriptor;
+  getFieldNames(): string[];
+  getApplyGlobalQuery(): boolean;
+  getIndexPatternIds(): string[];
+  getQueryableIndexPatternIds(): string[];
+  getGeoGridPrecision(zoom: number): number;
+  shouldBeIndexed(): boolean;
+  getPreIndexedShape(): Promise<PreIndexedShape | null>;
+  createFieldFormatter(field: IField): Promise<FieldFormatter | null>;
+  loadStylePropsMeta(args: unknown): Promise<unknown>; // todo
+  getValueSuggestions(field: IField, query: string): Promise<string[]>;
+}
+
+export class AbstractSource implements ISource {
+  static isIndexingSource = false;
+
+  readonly _descriptor: SourceDescriptor;
+  readonly _inspectorAdapters?: object;
+
+  constructor(descriptor: SourceDescriptor, inspectorAdapters?: object) {
+    this._descriptor = descriptor;
+    this._inspectorAdapters = inspectorAdapters;
+  }
+
+  destroy(): void {}
+
+  cloneDescriptor(): SourceDescriptor {
+    // @ts-ignore
+    return copyPersistentState(this._descriptor);
+  }
+
+  async supportsFitToBounds(): Promise<boolean> {
+    return true;
+  }
+
+  /**
+   * return list of immutable source properties.
+   * Immutable source properties are properties that can not be edited by the user.
+   */
+  async getImmutableProperties(): Promise<ImmutableSourceProperty[]> {
+    return [];
+  }
+
+  getInspectorAdapters(): object | undefined {
```

Comment:
Should we use [this](https://github.com/elastic/kibana/blob/39b233a24d966e346bc7676764904b1289a3e245/src/plugins/inspector/public/types.ts#L25) here?